servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/DelayedSubscription.java

```
@@ -18,84 +18,104 @@
     private static final AtomicLongFieldUpdater<DelayedSubscription> requestedUpdater =
-            AtomicLongFieldUpdater.newUpdater(DelayedSubscription.class, "requested");
+            newUpdater(DelayedSubscription.class, "requested");
 
-    @SuppressWarnings("unused")
     @Nullable
-    private volatile Subscription current;
-    @SuppressWarnings({"unused", "FieldCanBeLocal"})
+    private Subscription subscription;
     private volatile long requested;
 
     /**
      * Set the delayed {@link Subscription}. This method can only be called a single time and
      * subsequent calls will result in {@link #cancel()} being called on {@code delayedSubscription}.
      *
      * @param delayedSubscription The delayed {@link Subscription}.
+     * @return {@code true} if {@code delayedSubscription} was accepted as the delayed subscription. {@code false} if
+     * this method had already been called and {@code delayedSubscription} has been {@link #cancel() cancelled}.
      */
-    public void delayedSubscription(Subscription delayedSubscription) {
-        // Temporarily wrap in a ConcurrentSubscription to prevent concurrent invocation between this thread and
-        // a thread which may be interacting with this class's Subscription API.
-        final Subscription concurrentSubscription = wrap(delayedSubscription);
-        if (!currentUpdater.compareAndSet(this, null, concurrentSubscription)) {
-            delayedSubscription.cancel();
-        } else {
-            tryDrainRequested(concurrentSubscription);
-
-            // Unwrap the concurrent subscription because there will be no more concurrency.
-            currentUpdater.compareAndSet(this, concurrentSubscription, delayedSubscription);
+    public boolean delayedSubscription(Subscription delayedSubscription) {
+        requireNonNull(delayedSubscription);
+        for (;;) {
+            final long prevRequested = requested;
+            if (prevRequested <= INVALID_GREATEST_VALUE) {
```

I did a quick check on usages of this method and we always call it from `onSubscribe()`. Can we simply disallow calling this method twice? This will do two things:
- Fail fast for unexpected usage (we do not expect this to be called more than once)
- Simplify contract (returned `boolean` should always be `true`)

---

discussed offline. for now I'll just stick with `void` return type as `boolean` isn't strictly required. future considerations for throwing:
- more aggressive breaking change that could potentially break downstream `onSubscribe` propagation such that a subsequent `onError` is called without a prior `onSubscribe`:

```
onSubscribe(s) {
  mySubscription.delayedSubscription(s);
...
  downstreamSubscriber.onSubscribe(mySubscription);
}
onError(cause) {
  downstreamSubscriber.onError(cause); // onSubscribe wasn't called
}
```
- investigation required if `DelayedCancellable` (which is used in similar capacity for Single/Completable) should follow the same behavior (currently they are both `void`)